src/aspire/operators/diag_matrix.py

```diff
@@ -0,0 +1,622 @@
+
+        :return: A `DiagMatrix` like self.
+        """
+
+        return DiagMatrix(np.abs(self._data))
+
+    def __abs__(self):
+        """
+        Operator overload for absolute value of `DiagMatrix` instance.
+        """
+
+        return self.abs()
+
+    def pow(self, val):
+        """
+        Compute the elementwise power of `DiagMatrix` instance.
+
+        :param val: Value to exponentiate by.
+        :return: A `DiagMatrix` like self.
+        """
+
+        return DiagMatrix(self._data**val)
+
+    def __pow__(self, val):
+        """
+        Operator overload pow of `DiagMatrix` instance.
+        """
+
+        return self.pow(val)
+
+    @property
+    def norm(self):
+        """
+        Compute the norm of a `DiagMatrix` instance.
+
+        :return: The norm of the `DiagMatrix` instance.
+        """
+        # Elements of a diag matrix are its singular values,
+        #   and the norm is equal to the largest singular value.
+        return np.abs(self._data).max(axis=-1)
+
+    # Transpose methods are provided for reasons of interoperability
+    # in code that also uses with BlkDiagMatrix.
+    def transpose(self):
+        """
+        Get the transpose matrix of a `DiagMatrix` instance.
+
+        :return: The corresponding transpose form as a `DiagMatrix`.
+        """
+        return self.copy()
+
+    @property
+    def T(self):
+        """
+        Syntactic sugar for self.transpose().
+        """
+
+        return self.transpose()
+
+    def dense(self):
+        """
+        Convert `DiagMatrix` instance into full matrix.
+
+        :return: The `DiagMatrix` instance including the zero elements of
+        non-diagonal elements.
+        """
+
+        if self._data.ndim == 1:
+            return np.diag(self._data)
+
+        original_stack_shape = self.stack_shape
+        diags = self.stack_reshape(-1)  # Flatten stack
+
+        dense = []
+        for d in diags.asnumpy():
+            dense.append(np.diag(d))
+        # Convert to np.array
+        dense = np.concatenate(dense)
+        # Reshape to original stack shape.
+        return dense.reshape(*original_stack_shape, self.count, self.count)
+
+    def apply(self, X):
+        """
+        Define the apply option of a diagonal matrix with a matrix of
+        coefficient vectors.
+
+        :param X: Coefficient matrix, each column is a coefficient vector.
+        :return: A matrix with new coefficient vectors.
+        """
+
+        return self * X
+
+    def rapply(self, X):
+        """
+        Right apply.  Given a matrix of coefficient vectors,
+        applies the diagonal matrix on the right hand side.
+        Example, X @ self.
+
+        This is the right hand side equivalent to `apply`.
+
+        :param X: Coefficient matrix, each column is a coefficient vector.
+
+        :return: A matrix with new coefficient vectors.
+        """
+
+        #  Note there is an optimization opportunity here,
```

Comment:
Can you elaborate for posterity?

Reply:
I htink that was copy pasta from BlkDiagMatrix where I do some transposes, my bad. Removed comment.

Reply:
Ah. Good thing I gave up trying to figure it out after 2 minutes then…